pkg/sql/vecindex/high_speed_backfill_design.md

```diff
@@ -0,0 +1,1102 @@
+
+## Proposed Solution: Declarative Schema Changer Integration
+
+### Architecture Overview
+
+Rather than modifying the general backfill infrastructure, we propose adding a **new schema changer step** that pre-builds the vector index structure before the standard backfill phase. This keeps vector-specific logic isolated and follows existing schema changer patterns.
+
+### New Schema Changer Step: Structure Building
+
+The new `STRUCTURE_BUILDING` step performs one or more full scans over the primary index to construct the complete vector index hierarchy before any backfill operations begin. This approach:
+
+**Primary Index Scanning Strategy:**
+- **Initial Centroid Selection**: First scan uses random sampling (`randint() % 100 == 0`) to select ~1% of vectors as initial leaf partition centroids
+- **Batch-Based Refinement**: Subsequent scans process batches of 1000 random vectors, using mini-batch k-means to refine centroids until convergence
+- **Interior Partition Construction**: Final step builds interior partitions based on the optimized leaf centroids, handling prefixed indexes appropriately
+- **Shared Structure**: The resulting hierarchy is used by both the new vector index and any temporary indexes created during the process
+
+**Key Benefits:**
+- **Stable Partition Structure**: Enables effective caching of `SearchForInsert` results due to stable centroids
+- **Bulk Adder Compatibility**: Pre-built structure allows use of standard bulk adders (vastly faster than transactional inserts)
+- **Reduced Vector-Specific Code**: Minimizes vector-specific logic in general backfill and merge infrastructure
+- **Shared Structure Usage**: Both new and temporary indexes use the same optimized structure, avoiding metadata duplication
+- **Automatic Fixup Avoidance**: Fixup logic naturally avoids indexes under construction without explicit disabling
+- **Atomic Construction**: Structure building is atomic - either completes fully or rolls back cleanly
+
+### Enhanced Schema Changer Flow
+
+We introduce new statuses `STRUCTURE_BUILDING` and `STRUCTURE_BUILT` between `ABSENT` and `BACKFILL_ONLY`:
+
+```
+ABSENT → **STRUCTURE_BUILDING** → **STRUCTURE_BUILT** → BACKFILL_ONLY → BACKFILLED → DELETE_ONLY → MERGE_ONLY → MERGED → WRITE_ONLY → VALIDATED → PUBLIC
+```
+
+#### STRUCTURE_BUILDING Phase
+
+During this phase, the schema changer performs the computationally intensive work of building the optimal partition structure for the vector index:
+
+1. **Random Sampling Scan**: Performs an initial scan over the primary index, randomly selecting approximately 1% of vectors (`randint() % 100 == 0`) to serve as initial leaf partition centroids
+2. **Batch-Based Refinement**: Conducts subsequent scans that process batches of random vectors (typically 1000 per batch) to refine the initial centroids using mini-batch k-means until convergence
```

Comment:
Ideally, we would construct multiple 1000 random vector batches as part of a single scan over the primary index. That is, we'd scan all rows once, gathering 1% of vectors for initial centroids, and also constructing N batches, maybe using reservoir sampling to ensure randomness without knowing exactly how many rows there will be. However, one question to be resolved is what N should be. I assume we'd at least have an estimate of how many rows there are in the primary index; if not, maybe it's best to do 2 scans, 1 to get the initial centroids and another to assemble the random batches.